src/common.hpp

```diff
@@ -554,4 +554,35 @@ class Stopwatch
     std::chrono::high_resolution_clock::time_point tstart;
 };
 
+////////////////////////////////////////////////////////////////////////////////
+// A global log that can stream to an external source (like unreal).
+// Intended to replace calls to printf when stdout is not available.
+// Not thread safe or intended for speed.
```

Comment:
Yea, the logging got out of hand, and I should have just pulled it out - multiple things in one PR is bad form on my part, sorry. I can just pull it all out if you don't have time to work on it.
As for your specific comments:
1. The log is replacing printf (or maybe cout), both of which are pretty slow and have thread issues with interleaving. Logging warnings/errors is hopefully not a bottleneck. Compile flag explanation below in 3.
2. I agree it's thread safe if EXTERNAL_GLOBAL_LOG is, but that's set by the user. Right now, for example, if two threads write strings to the global log, it might mix or interleave the strings. The comment should be clarified.
3. I tested the GPU code by compiling and running the GPU version, and it generated bellhop output that looked fine. As you say, it won't output to the Unreal. Printf support in cuda appears to be strange (hence the compile flag) and vsprintf gave host/device compiler warnings and didn't compile. vprintf will compile with cuda, but it still gives the compiler warning. Cuda doesn't seem to support va, so the simple 'replace printf' approach is probably a dead end. Like I said, I should have just removed the hacked in log stuff and done this separately.